Add options to colorbar's title to set location, orientation and alignment

sebacastroh opened the pull request:
- [x] issues: fixes #14149
- [x] tests added / passed
- [x] release document entry (if new feature or API change)

As I exposed in #14149 , this PR add some optional arguments to the colorbar's title to define its location, orientation and alignment. I also modified the padding input so it can be a tuple with four values (top, right, bottom and left sides).

I'm using the [colorbar_log example](https://docs.bokeh.org/en/latest/docs/examples/basic/annotations/colorbar_log.html) as a test.

## Example 1
```python
color_bar = r.construct_color_bar(padding=(0, 0, 0, 10), title='Title', title_location='above')
plot.add_layout(color_bar, "right")
```
![imagen](https://github.com/user-attachments/assets/4f4a531f-e067-4265-8466-d8e1ac002620)

## Example 2

```python
color_bar = r.construct_color_bar(padding=(10, 0, 0, 0), title='Title', title_location='below', title_text_halign='center')
plot.add_layout(color_bar, "below")
```

![imagen](https://github.com/user-attachments/assets/3fa38634-aa25-4c34-82b7-1d1027138533)

## Example 3

```python
color_bar = r.construct_color_bar(padding=(0, 0, 0, 10), title='Title', title_location='right', title_orientation='horizontal', title_text_valign='middle')
plot.add_layout(color_bar, "right")
```

![imagen](https://github.com/user-attachments/assets/5c3e3c56-5931-4615-a0e8-d3a68db38bd8)

I'm still dealing with some combinations and the idea is to preserve the old behavior if the new arguments are omitted, so it's work in progress.

I'd like to read your opinions on this.

sebacastroh committed:
Add title_location and title_orientation options for colorbar (bokehjs)

sebacastroh committed:
Add title_location and title_orientation options for colorbar (bokehjs)

sebacastroh committed:
Add title_location and title_orientation options for colorba (bokeh)

sebacastroh committed:
Add title_location and title_orientation options for colorba (bokeh)

sebacastroh committed:
Add horizontal and vertical align options for colorbar title (bokehjs)

sebacastroh committed:
Add horizontal and vertical align options for colorbar title (bokehjs)

sebacastroh committed:
Add horizontal and vertical align options for colorbar title (bokeh)

sebacastroh committed:
Add horizontal and vertical align options for colorbar title (bokeh)

sebacastroh committed:
Expand padding input to receive values for top, right, bottom and left sides (bokehjs)

sebacastroh committed:
Expand padding input to receive values for top, right, bottom and left sides (bokehjs)

sebacastroh committed:
Expand padding input to receive values for top, right, bottom and left sides (bokeh)

sebacastroh committed:
Expand padding input to receive values for top, right, bottom and left sides (bokeh)

sebacastroh committed:
Fix missing parentheses

sebacastroh committed:
Fix missing parentheses

sebacastroh committed:
fix lint errors

sebacastroh committed:
fix lint errors

sebacastroh committed:
Add ticks before title when both are located right or below of the colorbar

sebacastroh committed:
Add ticks before title when both are located right or below of the colorbar

sebacastroh commented:
I'm trying to run the tests locally before to push my commits, but I'm getting the following error

```[11:23:15] Starting 'test:unit'...
Running in Chrome/130.0.6723.117 using devtools protocol 1.3
warning: 130.0.6723.117 is not supported; officially supported version is 118.0.5993.88
net::ERR_ABORTED
[11:23:17] Finished 'test:unit' after 1.80 s
[11:23:17] failed: tests failed
[11:23:17] failed: tests failed
```

I've been looking for that version of Chrome but I can't find it. I'm using Windows 11. Do you know where to find that version?

mattpap commented:
> I'm trying to run the tests locally before to push my commits, but I'm getting the following error (...)

See issue #14117. Currently testing bokehjs is only possible on Linux. You will have to rely on test output in CI. To resolve most of current test issues, run `pytest test/test_defaults.py` and commit the results. Also update bokeh's unit tests.

sebacastroh committed:
Add support for title location for Colorbar

sebacastroh committed:
Add support for title location for Colorbar

sebacastroh committed:
fix unit tests

sebacastroh committed:
fix unit tests

sebacastroh committed:
typo

sebacastroh committed:
typo

sebacastroh commented:
I think that I addressed all tests. There are 3 Windows' tests that failed, but it seems is not because of the changes I did 😬 .

## Discussion

I have added 4 new parameters to `ColorBar` to change the title location, orientation, horizontal and vertical alignment.

```python
title_orientation = String("auto", help="""
Wheter the color bar's title should be oriented upward,
downward or horizontal.
""")

title_location = Either(Enum(Location), Auto, default="auto", help="""
Specifies on which side of the color bar the title will be located.
""")

title_text_halign = Enum(HAlign, default="left", help="""
Specifies where to align color bar's title horizontally.
""")

title_text_valign = Enum(VerticalAlign, default="bottom", help="""
Specifies where to align color bar's title vertically.
""")
```

I've been thinking that maybe it could be reduced to 3 or even to 2. I'd like to read your opinions

### ```title_orientation```

This parameter allows you to write the title in different orientations: upward, downward and horizontal.

![imagen](https://github.com/user-attachments/assets/b2b54b3d-da8b-414e-88a8-451a4af0e2df)
_Figure 1: Example of the three available title orientations_

However, I'm not sure if this function is really neccessary, and maybe it is better to use an automatic behaviour. Currently, the title orientation behaves as follows when is omitted:
- If the colorbar is horizontal, the title is horizontal regardless of the title's location.
- If the color bar is vertical and the title's position is above or below it, the title will be horizontal. However, if the title is positioned to the left of the color bar, it will be oriented upward, and if it is to the right, it will be oriented downward.

![imagen](https://github.com/user-attachments/assets/256f48b7-ca50-4a71-8b20-cd67465e49a5)
_Figure 2: Automatic behavior of the title orientation depeding on the colorbar position and title location_

On the other hand, I'm using two values that are not listed in the typical enumerations used in Bokeh (upward and downward), I don't know if we should add these values for something that specific. I'm inclined to use always the automatic behavior. What do you think?

### ```title_text_halign``` and ```title_text_valign```

I added these two parameters because I noticed that the original parameter ```title_text_align``` is not being passed properly to the `Title` object 

```diff
  protected _apply_title_properties(): void {
    const attrs: Partial<Title.Attrs> = {
      text: this.model.title ?? "",
      standoff: this.model.title_standoff,
+     align: this.model.title_text_halign,
+     vertical_align: this.model.title_text_valign,
      // TODO: this needs strict typing
      ...mixins.attrs_of(this.model, "title_", mixins.Text, false),
    }
    this._title.setv(attrs)
  }
```
I see two options:
1. We could deprecate the ```title_text_align``` parameter and mantain the explicit horizontal and vertical alignment, or
2. We could change (fix) the ```title_text_align``` parameter in order to capture the horizontal and vertical alignment in one single parameter. That would require, I think, to use a definition like "above = left", "middle = center" and "below = right".

## Change padding parameter

In addition, I changed the ```padding``` parameter so it can be a single value (as it is now) or a tuple of four numbers, so the user can indicate different padding values for the top, right, bottom and left sides.

## Default behavior

The default values were chosen in order to maintain the current behavior, so the users should not see any difference if they update the Bokeh library.

Again, I'd like to know your thoughs. As @mattpap mentioned in #14149 this could be implemented for the `ScaleBar` and maybe the `Legend` models, but I think it is better to discuss the behavior before to implement it in other models.

mosc9575 commented:
This PR solves a problem I have right now. What is the reason this is not merged? Maybe I can help to bring this to an end.

mattpap committed:
Fix types and add *.pyi type declarations

mattpap committed:
Fix types and add *.pyi type declarations

mattpap commented:
> What is the reason this is not merged? Maybe I can help to bring this to an end.

I guess it got lost in a never ending stream of PRs. I rebased this on top of `branch-3.10` and did some cleanup of the types. What we really need is some visual integration tests. If you could help with that, we could get this merged pretty soon.

mosc9575 commented:
I tried to add this [commit](https://github.com/sebacastroh/add_title_location_option_in_colorbar/commit/89d1147678e91ba948f189d3e46d8ef5c03935e3) to the branch. I did something wrong and have to figure out later how to fix this.

bryevdv commented:
I want to try and salvage this PR but given how long it has been open I am just going to update the branch and try to push some small fixes and test directly

bryevdv committed:
Merge branch 'branch-3.10' into 14149_title_colorbar

sebacastroh commented:
I'm glad this PR is still getting attention! Unfortunately, I don't have time to help with this right now, but I really hope it can be merged if you find it useful, thanks for all your help!

bryevdv committed:
Fix color bar title positioning

bryevdv committed:
Update color bar visual baselines

bryevdv committed:
Clarify color bar alignment visual tests

bryevdv committed:
Update clarified color bar visual baselines

bryevdv committed:
Make asymmetric color bar padding evident

bryevdv committed:
Update evident padding visual baseline

bryevdv committed:
Remove redundant color bar alignment cases

bryevdv committed:
Update concise alignment visual baseline

bryevdv commented:
OK, I think this is as reasonably ready as it is going to get

bryevdv commented:
@philippjfr any thoughts on this?

bryevdv commented:
Thanks @sebacastroh 

I've added `NEEDS BACK PORT` in case we do release a 3.11 before 4.0
